pallene/to_ir.lua

```diff
@@ -87,26 +106,44 @@ function ToIR:exit_function(cmds)
     end
 end
 
---
---
---
-function ToIR:is_local(decl)
-    if decl._tag == "ast.Decl.Decl" then
-        local loc_id = self.loc_id_of_decl[decl]
-        local glb_id = self.glb_id_of_decl[decl]
+-- Returns a to_ir.Var object containing ID of `decl`.
+function ToIR:resolve_variable(decl)
+    if decl._tag == "ast.Var.Name" then
+        local glb_id = assert(self.glb_id_of_var[decl])
+        return to_ir.Var.GlobalVar(glb_id)
+    end
+    assert(decl._tag == "ast.Decl.Decl")
+
+    local glb_id = self.glb_id_of_decl[decl]
+    if glb_id then
+        return to_ir.Var.GlobalVar(glb_id)
+    end
+
+    local var
+    local i = 1
+    while i <= #self.func_stack do
+        local loc_id = self.func_stack[i].loc_id_of_decl[decl]
         if loc_id then
-            assert(not glb_id)
-            return true, loc_id
+            var = to_ir.Var.LocalVar(loc_id)
+            break
+        end
+        i = i + 1
+    end
```

Comment:
Style suggestion:

```
local stack_id, var
for i, func in ipairs(self.func_stack)
    loc_id = ...
    if loc_id then
        stack_id = i
        var = ...
        break
    end
end
```

It's a matter of taste, but this lets us get rid of the mutable `i` variable and would allow us to rename the `j` so we do not have an `i` and `j` at the same time.